### DISCUSSION TITLE
Access remote events from dataviewjs

### DISCUSSION THREAD

**Discussion by renarl** [davish/obsidian-full-calendar#394] [Feature Requests]
I see an example of how to call calendar view from dataviewjs. Is there also a way to get a list of event data from dataviewjs so that I can show them in a table?

**Comment by gala8y** [comment]
Hi, this little snippet shows table with events for any given day.
Current file (where you put this code into) has to have a date in file name (for example, it can be your daily note).

```dataview
table startTime as "start @", endTime as "end @",  date(dateformat(file.day, "yyyy-MM-dd") + "T" + endTime) - date(dateformat(file.day, "yyyy-MM-dd") + "T" + startTime) as duration
from "folder name where your events are"
where file.day = this.file.day
sort startTime asc
```

**Comment by renarl** [comment]
Thanks for the suggestion. It works partially - I can get the local events I've created. But my primary need is to get remote events from iCloud and Google calendars so I can optionally make local event notes for some of them.

PS here is a `dataviewjs` analog for getting local events if someone needs it:
````
```dataviewjs
// !! assuming that your file name is a date
const currFileName = dv.current().file.name;
const currentDay = luxon.DateTime.fromISO(currFileName);

// !! change this to the folder name where you revents are  
const eventsFolderPath = "folder name where your events are";


const pagesThatRefferToToday = dv.pages(`"${eventsFolderPath}"`)
  .where(p => (
    p.date.equals(currentDay)
  ))
  .flatMap(p => [
    dv.fileLink(`${eventsFolderPath}/${p.file.name}`, false, `${p.title} - ${p.startTime} to ${p.endTime}`)
]);

dv.list(
  pagesThatRefferToToday
)
```
````

**Comment by davish** [comment]
This is an interesting feature request, thanks @renarl!

**Comment by KosmosisDire** [comment]
I second this! I am trying to get info about remote .ical events using dataviewjs. It would be amazing to have this functionality.

**Comment by neoromantic** [comment]
Came here to read about this particular feature, whether it's being discussed.

I cannot consider calendar functionality in Obsidian complete, if I can't integrate remote calendar events  (from my iCloud or Google) in my daily notes.

I can do it with other plugins, I guess, but since we already have sync here, it'd be great if I could just query calendar events similar to how I query various tasks now.

**Comment by careilly** [comment]
I cooked this horror up, which list events for today, but does not include repeating events. `getAllEvents()` doesn't give them - I assume there's something clever I have to do to expand the recurrence rules, but I don't need them for my use case.

```
```dataviewjs
let cache = app.plugins.plugins["obsidian-full-calendar"].cache
var result= []
cache.getAllEvents().forEach((x, i) => {
	x.events.forEach((e)=> {
		var event = e.event
		var date = dv.date(event.date)
		var filedate = dv.date(dv.current().file.cday)
		if (date && date.day==filedate.day && date.month==filedate.month && date.year == filedate.year ) {
		    result.push(event.startTime+"-"+event.endTime+" "+event.title)
		} 
	})
})	

if (result.length==0) {
	dv.el("b","No calendar entries today.")
} {
	dv.list(result)
}
```
```

**Reply by careilly** [reply]
If FullCalendar exposed a version of the rrule expand function that parsed the rules and expands occurrences over a range it might do the job. `expandRecurrances(event, start,end)` maybe?

**Reply by careilly** [reply]
I made this work, but I had to write a plugin to import and expose the bits of the rrule package I needed! @davish is that something you'd want to pull into this plugin or something I should just distribute myself?

I guess I'll use it for a week or two and then package something up. It's possibly too hacky to expose right now! 

On the up side, I learned how to write a plugin.

**Comment by mathisgauthey** [comment]
I'm looking at this particular feature as well for ma daily-note template. Any update ?

**Comment by erdalgunes** [comment]
I've iterated on @careilly's [solution](https://github.com/davish/obsidian-full-calendar/discussions/394#discussioncomment-6427949) a bit. I've noticed the recurring events from google calendar have a type of `rrule` and property `startDate` instead of `date`. I also added a logic for the recurring events of the plugin.

I don't know if it covers all the cases out there but it works for my calendar. You can improve the filter according to your needs.

~~~
```dataviewjs
// Extracting Obsidian Full Calendar cache at once.
let plugin = app.plugins.plugins["obsidian-full-calendar"];
let cache = plugin ? plugin.cache : null;
if (!cache) {
  throw new Error("Plugin not found or Cache is empty.");
}

// Getting current file date.
let filedate = dv.date(dv.current().file.cday);
let fileDayOfWeek = filedate.day; // 0 is Sunday, 1 is Monday, ..., 6 is Saturday

// A mapping of day character to corresponding integer
const dayOfWeekMap = {
  0 : 'U',
  1 : 'M',
  2 : 'T',
  3 : 'W',
  4 : 'R',
  5 : 'F',
  6 : 'S'
};

let allEvents = cache.getAllEvents();

function filterFunction(e) {
  if (e.event.type === "recurring") {
    return handleRecurringEvent(e, filedate, fileDayOfWeek);
  }

  if (e.event.type === "rrule") {
    return handleRRuleEvent(e, filedate);
  }
  
  return handleNormalEvent(e, filedate);
}

let result = allEvents
  .flatMap(x => x.events)
  .filter(filterFunction)
  .map(e => `${e.event.startTime}: ${e.event.title}`);

// If no events are found for the current file day, display a message.
if (result.length === 0) {
  dv.el("b", "No calendar entries today.");
} else {
  dv.list(result);
}

function handleRecurringEvent(e, filedate, fileDayOfWeek) {
  // Check if the event repeats on the current day
  let isInDayOfWeek = e.event.daysOfWeek.includes(dayOfWeekMap[fileDayOfWeek]);
  // If startRecur exists, check if the current date falls within this range
  let isInDateRange = true;
  if (e.event.startRecur) {
    let startRecur = new Date(e.event.startRecur);
    let endRecur = e.event.endRecur ? new Date(e.event.endRecur) : new Date(9999, 12, 31);

    isInDateRange = (filedate >= startRecur) && (filedate <= endRecur);
  }
  return isInDayOfWeek && isInDateRange;
}

function handleRRuleEvent(e, filedate) {
  let date = dv.startDate(e.event.date);
  return date.day === filedate.day && date.month === filedate.month && date.year === filedate.year;
}

function handleNormalEvent(e, filedate) {
  let date = dv.date(e.event.date);
  return date.day === filedate.day && date.month === filedate.month && date.year === filedate.year;
}
```
~~~
